contrib/nginx.md

```diff
@@ -0,0 +1,106 @@
+
+## http server: websocket proxying
+
+First, configure Soju to listen for websocket on the `/run/soju/ws.sock` unix domain socket:
+```
+…
+listen ws+unix:///run/soju/ws.sock
+…
+```
+
+Then configure nginx to proxy all http traffic for the `/socket` prefix to that unix domain socket,
+by using the `proxy_pass` directives:
+
+```
+# /etc/nginx/sites-enabled/60-soju.conf
+server {
+  server_name chat.example.com;
+  listen 80;
+  listen [::]:80;
+
+  location /socket {
+    proxy_pass http://unix:/run/soju/ws.sock:/socket;
+    proxy_read_timeout 600s;
+    proxy_http_version 1.1;
+    proxy_set_header Upgrade $http_upgrade;
+    proxy_set_header Connection "Upgrade";
+    proxy_set_header X-Forwarded-For $remote_addr;
+    proxy_set_header X-Forwarded-Proto $scheme;
+    proxy_set_header Host $host;
+  }
+
+  root /srv/gamja/;
+}
+```
+All the rest of http requests will be served from staic files in the `/src/gamja` directory
+(you can just copy the gamja `dist/` production files there).
+
+Once you have the above configuration, you can run `sudo certbot --nginx -d chat.example.com`
+and you'll get the configuration file updated for TLS, redirects, and some security headers.
+
+
+## stream server: irc protocol proxying
+
+Nginx, since version 1.9.0, supports the [stream](https://nginx.org/en/docs/stream/ngx_stream_core_module.html)
+module, which can proxy non-http protocols. We can use that to make nginx accepts/handle the tls protocol
+for the irc tcp connections.
+
+Configure Soju to listen on the unix socket `/run/soju/irc.sock` for irc connections:
+```
+…
+listen unix:///run/soju/irc.sock
+…
+```
+
+Now configure nginx to listen on the 1667 tcp port (an an example), and proxy to soju using the "PROXY protocol".
+The `ssl_certificate`, `ssl_certificate_key` and `ssl_dhparam` are copy/pasted from the http server config
+(as edited by certbot).
+```
+# /etc/nginx/streams-enabled/60-soju.conf
+server {
+  listen      1667 ssl;
+  listen      [::]:1667 ssl;
+  proxy_pass unix:/run/soju/irc.sock;
+  proxy_protocol on;
+
+  ssl_certificate /etc/letsencrypt/live/chat.example.com/fullchain.pem;
+  ssl_certificate_key /etc/letsencrypt/live/chat.example.com/privkey.pem;
+  ssl_dhparam /etc/letsencrypt/ssl-dhparams.pem;
+}
+```
+
+
+## Socket file permissions
+
+The `/run/soju/ws.sock` and `/run/soju/irc.sock` unix socket pseudo-files are created by soju,
+but need to be accessible by nginx. A common solution to that issue is to run soju with the same
+group as nginx is running (for example www-data in debian). In that case the unix socket file permissions
+should be something like:
+
+```
+drwxr-xr-x 2 soju www-data 100 May  1 03:00 /run/soju/
+srwxrwxr-x 1 soju www-data 0 May  1 03:00 /run/soju/irc.sock
+srwxrwxr-x 1 soju www-data 0 May  1 03:00 /run/soju/ws.sock
+```
+A common way to establish that when running soju under debian/ubuntu is to use the `Group=` directive
+in the soju .service unit file (or a drop-in):
+```
+[Service]
+Group=www-data
```

Comment:
Maybe `SupplementaryGroups` would be better suited, so that the primary group isn't erased?

Hm, I suppose this may not be enough? Would it work to add a drop-in for nginx to add the soju supplementary group?

Reply:
> Maybe `SupplementaryGroups` would be better suited, so that the primary group isn't erased?

it can work like that, but then `soju` would need to explicitly set the group of the unix socket files.

> Hm, I suppose this may not be enough? Would it work to add a drop-in for nginx to add the soju supplementary group?

The nginx service runs as root and explicitly switches user/group by itself. not sure it supports supplementary groups either :/